### PR TITLE
added support for fst optimization in the log semiring

### PR DESCRIPTION
**Pull request by riccardobassanisonos**

### CODE REVIEW

File: rustfst-ffi/src/algorithms/optimize.rs [hunk 1]
```diff
@@ -18,3 +20,24 @@ pub extern "C" fn fst_optimize(ptr: *mut CFst) -> RUSTFST_FFI_RESULT {
         Ok(())
     })
 }
+
+use ffi_convert::RawPointerConverter;
```
Comment (Garvys) [inline rustfst-ffi/src/algorithms/optimize.rs:24]: Move this import at the top of the file

File: rustfst-python/rustfst/algorithms/optimize.py [hunk 2]
```diff
@@ -13,7 +16,17 @@ def optimize(fst: VectorFst):
     Args:
         fst: Fst to optimize.
     """
-
     ret_code = lib.fst_optimize(fst.ptr)
     err_msg = "Error during optimize"
     check_ffi_error(ret_code, err_msg)
+
+
+def log_optimize(fst: VectorFst):
```
Comment (Garvys) [inline rustfst-python/rustfst/algorithms/optimize.py:24]: Regarding naming, i think `optimize_in_log` would be more meaningfull.

File: rustfst-python/tests/algorithms/test_optimize.py [hunk 3]
```diff
@@ -55,3 +55,35 @@ def test_optimize_fst():
...
+    tr1_3 = Tr(0, 0, 3.0, s4)
+    fst1.add_tr(s2, tr1_3)
+
+    tr1_4 = Tr(4, 6, 4.0, s4)
+    fst1.add_tr(s2, tr1_4)
+
+    tr1_5 = Tr(7, 8, 5.0, s4)
+    fst1.add_tr(s3, tr1_5)
+
+    fst1.optimize()
```
Comment (Garvys) [inline rustfst-python/tests/algorithms/test_optimize.py:87]: Are sure you're testing the correct function ? ^^

**Review by Garvys — CHANGES_REQUESTED** [review]

File: rustfst-ffi/src/algorithms/optimize.rs [hunk 4]
```diff
@@ -18,3 +21,23 @@ pub extern "C" fn fst_optimize(ptr: *mut CFst) -> RUSTFST_FFI_RESULT {
         Ok(())
     })
 }
+
+#[no_mangle]
+pub extern "C" fn fst_optimize_in_log(ptr: *mut *const CFst) -> RUSTFST_FFI_RESULT {
+    wrap(|| {
+        let lol = unsafe { *ptr };
```
[the remark this answers is not in the document]
Comment (riccardobassanisonos) [reply rustfst-ffi/src/algorithms/optimize.rs:28]: ahah I thought we had changed it good catch

**Review by Garvys — APPROVED** [review]
Well done !

**Pull request merged**